gms/gossiper.cc

```diff
@@ -2105,10 +2105,18 @@ future<> gossiper::add_saved_endpoint(inet_address ep, permit_id pid) {
     auto ep_state = endpoint_state();
     auto es = get_endpoint_state_ptr(ep);
     if (es) {
+        if (es->get_heart_beat_state().get_generation()) {
```

Comment:
As I wrote in https://github.com/scylladb/scylladb/issues/16429

To me it seems that nothing has actually changed.

First, the commit message suggests that the existing code "updates its heart_beat_state".

But assuming the assertion you put here is indeed true, i.e. add_saved_endpoint is only called when generation == 0, then the existing code does nothing. Indeed, it sets heart_beat_state to `heat_beat_state()` i.e. generation = 0, version = 0. But if generation == 0, that should actually not change anything.

So the only thing this code seems to do, is to actually add the assertion, not change any behavior.

Reply:
There is a subtle change where `topology_state_load` acquires the endpoint lock for checking if the endpoint state already exists.  Previously, it did so without acquiring the lock and so it could race with `add_saved_endpoint` that could have found such a state after acquiring the lock, and that state could have non-zero generation.

That's the gist of the fix.

Reply:
Ok, that makes sense, but can you please update the PR cover letter and commit messages (last one at least) so they explain this? I.e. the fact that we actually never have to "keep existing heart_beat_state" (other than 0), there was the possibility that it was non-zero when we called add_save_endpoint, but that was a bug which this PR fixes (second commit, as I understand) and now adds an assertion to prevent the bug from reoccuring.

Reply:
Also...

> it could race with add_saved_endpoint

so both topology_state_load and **some other place** would concurrently call `add_saved_endpoint`
what is that other place?

And shouldn't we ensure that this other place is always called first (e.g. before we start Raft group 0 server), instead of introducing locks?
cc @gusev-p

Reply:
No, the race is with `do_apply_state_locally`.

Reply:
> AFAIU topology_state_load is called in normal operation mode, not only when raft group is starts.

Yes, it's called on every topology change, for example.

> No, the race is with do_apply_state_locally.

So wait. But that means `add_saved_endpoint`, the one that is being called from `topology_state_load`, **will** observe non-zero generations, no? Because gossiping is happening in the background and is updating endpoint states received from nodes and setting non-zero generations from them. And then when we perform a topology change, `topology_state_load` calls `add_saved_endpoint` and your assertion should get triggered, no?

Now I'm even more confused -- why the tests are passing in CI...

>  That's why I think calling add_saved_endpoint there is an abuse of the interface.

I agree and @gusev-p also mentioned in another comment he doesn't like it.

I think in raft-topology mode, `add_saved_endpoint` should only be called once for each known peer **before** we start group 0 server (and outside `topology_state_load`), and its only purpose would be to restore known IP mappings from before the node stopped. Once that is done, we should never be calling `add_saved_endpoint`.

Reply:
>  And then when we perform a topology change, topology_state_load calls add_saved_endpoint and your assertion should get triggered, no?

`add_saved_endpoint` is only called for non-existing IP-s, which should have zero generation/version. The race this PR fixes is that `do_apply_state_locally` might have been called after the if in `topology_state_load` which checks `endpoint_state` doesn't exist.

Reply:
> I think in raft-topology mode, add_saved_endpoint should only be called once for each known peer before we start group 0 server (and outside topology_state_load), and its only purpose would be to restore known IP mappings from before the node stopped. Once that is done, we should never be calling add_saved_endpoint.

The reason I added `add_saved_endpoint` into `topology_state_load` is that it's the place where the current raft topology become known. We can't use `system.topology` before group0 start because it may be stale.

Reply:
Ok, so how about my proposal -- calling it before we start group 0 server, based on contents of system.peers table? Exactly like we do for gossip-based topology changes? Nodes that are appearing during normal operation (not when we restart) are not "saved endpoints", to it seems weird to call add_saved_endpoint from inside topology_state_load which is used during normal operation. If I understand the purpose of add_saved_endpoint, it's for restoring gossiper state on restart that we had before restart.

Reply:
In the future, I think we should consider a refactoring of the endpoint states map and allow updating it both from gossip and from raft.  In that model we may want a heart_beat_version per application state rather than per endpoint state.

This was raft can manage the topology-related states while gossip can update auxiliary states like the heat map, large partitions, etc.

Reply:
> The reason I added add_saved_endpoint into topology_state_load is that it's the place where the current raft topology become known. We can't use system.topology before group0 start because it may be stale.

Ah, but gossiper state that is supposed to be restored by add_saved_endpoint is not persisted in system.topology, it's persisted in system.peers.

And we don't care about the fact that it's stale, we're restarting and it was stale the same way as it was before we restarted. Further gossiping will update it.

Reply:
Or we can rely on class topology to hold all the critical stuff and it will be the source of truth.

Reply:
> And we don't care about the fact that it's stale, we're restarting and it was stale the same way as it was before we restarted. Further gossiping will update it.

Is it guaranteed that all the endpoints we see in raft topology will somehow eventually appear in gossiper? I see that we `remove_endpoint`-s from gossiper in `topology_state_load`, but we don't add them except through `add_saved_endpoint`.

And why not use the `system.topology` instead of `system.peers` than? `system.peers` is considered legacy.

Reply:
> Is it guaranteed that all the endpoints we see in raft topology will somehow eventually appear in gossiper?

Nodes are gossiping on their own. If a node is in raft topology and is alive, then it's gossiping with other nodes in raft topology. So they should see each other.

> I see that we remove_endpoint-s from gossiper in topology_state_load, but we don't add them except through add_saved_endpoint.

We never carefully designed how the interaction between raft-topology and gossiper should look like... Perhaps we should do it now.

I think that the only things that raft-topology should do to gossiper are:
- to tell it, through host-ID based interface, which nodes are banned. Gossiper remove all information about and would not further interact with nodes whose host IDs are banned
- update certain application states sometimes (tokens, status). But this could be done only for local node, no need to do it for peers. Nodes would update peers' information in gossiper through regular gossiping.

> And why not use the system.topology instead of system.peers then? system.peers is considered legacy.

In raft-topology mode, system.peers still has a single role that is required (but not legacy) -- persisting IP mappings and restoring them on boot. And that is also the only thing that we would like to restore into gossiper. system.topology doesn't have mappings.

Reply:
> If a node is in raft topology and is alive, then it's gossiping with other nodes in raft topology.

To be precise.. it's gossiping with nodes that it can reach through transitive closure obtained from seeds graph. But those are the only nodes it can gossip with --- even if raft topology contains information about certain node that is not in this transitive closure, we won't be able to gossip with it, because the transitive closure is the only way we can obtain the node's IP, which is needed to gossip with it.

**Raft topology can only restrict the set of nodes we're gossiping with, not extend it**

Reply:
FYI, one more site that I'd like to (ab)use add_saved_endpoint to inject an endpoint state, as a followup, is in prepare_replacement_info for ignored dead nodes.  This is required since the replacing node as no system.peers yet and those dead nodes qre visbile after full-cluster restart only in the gossip shadow round because they are not actively gossiping.

See https://github.com/scylladb/scylladb/pull/15788

Reply:
I propose we move this discussion somewhere else.
I created a new issue where I propose a solution. https://github.com/scylladb/scylladb/issues/16644

Reply:
> I propose we move this discussion somewhere else. I created a new issue where I propose a solution. #16644

Thanks.  I suggest going with this change for the time being and then look for improvements as follow up.